pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -256,38 +257,63 @@ func (h *upgradeHandler) OnRemove(_ string, upgrade *harvesterv1.Upgrade) (*harv
 	}
 
 	logrus.Debugf("Deleting upgrade %s", upgrade.Name)
+	return upgrade, h.cleanup(upgrade, true)
+}
+
+func (h *upgradeHandler) cleanup(upgrade *harvesterv1.Upgrade, cleanJobs bool) error {
+	// delete vm and images
+	repo := NewUpgradeRepo(h.ctx, upgrade, h)
```

Comment:
can we set repo to the handler instead.

Reply:
Hi Guangbo, do you mean to set it as a parameter of the function or a field of `upgradeHandler` structure? Thank you.

Reply:
Let's improve this in another following PR due to time constraints.

Reply:
https://github.com/harvester/harvester/issues/2939